src/html-imports.js

```diff
@@ -9,718 +9,754 @@
+  /**
+   * Importer will:
+   * - load any linked import documents (with deduping)
+   * - whenever an import is loaded, prompt the parser to try to parse
+   * - observe imported documents for new elements (these are handled via the
+   *   dynamic importer)
+   */
+  class Importer {
+    /**
+     * @param {!HTMLDocument} doc
+     */
+    constructor(doc) {
+      this.documents = {};
+      this._doc = doc;
+      // Make sure to catch any imports that are in the process of loading
+      // when this script is run.
+      const imports = doc.querySelectorAll(IMPORT_SELECTOR);
+      for (let i = 0, l = imports.length; i < l; i++) {
+        whenElementLoaded(imports[i]);
+      }
+      // Observe only document head
+      new MutationObserver(this._onMutation.bind(this)).observe(doc.head, {
+        childList: true
+      });
+
+      if (!useNative) {
+        this._loader = new Loader(
+          this._onLoaded.bind(this), this._onLoadedAll.bind(this)
+        );
+        whenDocumentReady(doc).then(() => this._loadSubtree(doc));
+      }
+    }
+
+    _loadSubtree(doc) {
+      const nodes = doc.querySelectorAll(IMPORT_SELECTOR);
+      // add these nodes to loader's queue
+      this._loader.addNodes(nodes);
+    }
+
+    _onLoaded(url, elt, resource, err, redirectedUrl) {
+      flags.log && console.log('loaded', url, elt);
+      // We've already seen a document at this url, return.
+      if (this.documents[url] !== undefined) {
+        return;
+      }
+      if (err) {
+        this.documents[url] = null;
+      } else {
+        // Generate an HTMLDocument from data.
+        const doc = makeDocument(resource, redirectedUrl || url);
+        // note, we cannot use MO to detect parsed nodes because
+        // SD polyfill does not report these as mutations.
+        this._loadSubtree(doc);
         this.documents[url] = doc;
       }
-      // don't store import record until we're actually loaded
-      // store document resource
-      elt.__doc = doc;
-    }
-  },
-
-  bootDocument: function(doc) {
-    this.loadSubtree(doc);
-  },
-
-  loadedAll: function() {
-    this._flatten(document);
-    runScripts();
-    this._fireEvents(document);
-    this.observe(document.head);
-  },
-
-  _flatten: function(element) {
-    var n$ = element.querySelectorAll(IMPORT_SELECTOR);
-    for (var i=0; i < n$.length; i++) {
-      var n = n$[i];
-      n.import = this.documents[n.href];
-      if (n.import && !n.import.__firstImport) {
-        n.import.__firstImport = n;
-        this._flatten(n.import);
-        if (!n.import.parentNode) {
-          n.appendChild(n.import);
-          if (document.contains(n.parentNode)) {
-            // TODO(sorvell): need to coordinate with observer in document.head.
-            //this.observe(n.import);
+    }
+
+    _onLoadedAll() {
+      this._flatten(this._doc);
+      Promise.all([
+        runScripts(this._doc),
+        waitForStyles(this._doc)
+      ]).then(() => fireEvents(this._doc));
```

Comment:
I believe it's safer to wait for the styles to load then execute the scripts, and finally fire the events. Will make it slower but more correct